### ISSUE TITLE
OSX

### ISSUE THREAD

**Issue by dmaynor** [sgadrat/pping#1]
davidmaynor@MacBook-Air pping % make pping
cc     pping.c   -o pping
In file included from pping.c:111:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:92:4: error: unknown type name 'n_short'; did you mean 'u_short'?
                        n_short icd_id;
                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_short.h:30:25: note: 'u_short' declared here
typedef unsigned short  u_short;
                        ^
In file included from pping.c:111:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:93:4: error: unknown type name 'n_short'; did you mean 'u_short'?
                        n_short icd_seq;
                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_short.h:30:25: note: 'u_short' declared here
typedef unsigned short  u_short;
                        ^
In file included from pping.c:111:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:99:4: error: unknown type name 'n_short'; did you mean 'u_short'?
                        n_short ipm_void;
                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_short.h:30:25: note: 'u_short' declared here
typedef unsigned short  u_short;
                        ^
In file included from pping.c:111:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:100:4: error: unknown type name 'n_short'; did you mean
      'u_short'?
                        n_short ipm_nextmtu;
                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_short.h:30:25: note: 'u_short' declared here
typedef unsigned short  u_short;
                        ^
In file included from pping.c:111:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:121:4: error: unknown type name 'n_time'
                        n_time its_otime;
                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:122:4: error: unknown type name 'n_time'
                        n_time its_rtime;
                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:123:4: error: unknown type name 'n_time'
                        n_time its_ttime;
                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:126:14: error: field has incomplete type 'struct ip'
                        struct ip idi_ip;
                                  ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet/ip_icmp.h:126:11: note: forward declaration of 'struct ip'
                        struct ip idi_ip;
                               ^
pping.c:143:17: error: field has incomplete type 'struct icmphdr'
        struct icmphdr hdr;
                       ^
pping.c:143:9: note: forward declaration of 'struct icmphdr'
        struct icmphdr hdr;
               ^
pping.c:144:22: error: invalid application of 'sizeof' to an incomplete type 'struct icmphdr'
        char msg[PING_PKT_S-sizeof(struct icmphdr)];
                            ^     ~~~~~~~~~~~~~~~~
pping.c:143:9: note: forward declaration of 'struct icmphdr'
        struct icmphdr hdr;
               ^
pping.c:246:30: error: use of undeclared identifier 'SOL_IP'
        if (setsockopt(ping_sockfd, SOL_IP, IP_TTL, net_payload(&ttl_val), sizeof(ttl_val)) != 0) {
                                    ^
11 errors generated.
make: *** [pping] Error 1
davidmaynor@MacBook-Air pping %

**Comment by sgadrat** [comment]
Hey, thanks for the report!

Could you try to change the line `#include <netinet/in.h>` to `#include <netinet/ip.h>` in `pping/pping.c`?

Also, you should compile `pping_sample.c`, `pping.c` is the library's implementation and has no `main()`. As I did not added any makefile, the compilation line should be `cc pping_sample.c pping/pping.c -o pping_sample` at the root of this repository.

**Comment by dmaynor** [comment]
Major improvement. Also I figured I would build the sample if the lib worked. I might be outdated with using make without makefiles. 

davidmaynor@MacBook-Air pping % make pping
cc     pping.c   -o pping
pping.c:143:17: error: field has incomplete type 'struct icmphdr'
        struct icmphdr hdr;
                       ^
pping.c:143:9: note: forward declaration of 'struct icmphdr'
        struct icmphdr hdr;
               ^
pping.c:144:22: error: invalid application of 'sizeof' to an incomplete type 'struct icmphdr'
        char msg[PING_PKT_S-sizeof(struct icmphdr)];
                            ^     ~~~~~~~~~~~~~~~~
pping.c:143:9: note: forward declaration of 'struct icmphdr'
        struct icmphdr hdr;
               ^
pping.c:246:30: error: use of undeclared identifier 'SOL_IP'
        if (setsockopt(ping_sockfd, SOL_IP, IP_TTL, net_payload(&ttl_val), sizeof(ttl_val)) != 0) {
                                    ^
3 errors generated.

**Comment by sgadrat** [comment]
Nice, this patch should do the trick:

```
diff --git a/pping/pping.c b/pping/pping.c
index ac6b08c..71022e8 100644
--- a/pping/pping.c
+++ b/pping/pping.c
@@ -103,8 +103,40 @@
        }
 
 #else
+
+#ifdef __APPLE__
+
+       // If the OS doesn't declare it, do it ourself (copy-pasted from GNU C Library, license: LGPL)
+#      include <stdint.h>
+       struct icmphdr
+       {
+               uint8_t type;           /* message type */
+               uint8_t code;           /* type sub-code */
+               uint16_t checksum;
+               union
+               {
+                       struct
+                       {
+                               uint16_t        id;
+                               uint16_t        sequence;
+                       } echo;                 /* echo datagram */
+                       uint32_t        gateway;        /* gateway address */
+                       struct
+                       {
+                               uint16_t        __unused;
+                               uint16_t        mtu;
+                       } frag;                 /* path mtu discovery */
+                       /*uint8_t reserved[4];*/
+               } un;
+       };
+
+       // Fix slightly changed names
+#      define SOL_IP IPPROTO_IP
+
+#endif
+
 #      include <sys/socket.h>
-#      include <netinet/in.h>
+#      include <netinet/ip.h>
 #      include <arpa/inet.h>
 #      include <netdb.h>
 #      include <unistd.h>
```

**Comment by sgadrat** [comment]
Fixed by 9aff06bb374b4ade8b6f51734b17f73d86783955
